### PR TITLE
Migrate current integration batch tests to equivalent MSQ tests

### PR DESCRIPTION
**Pull request by abhagraw**
### Description

Migrate current integration batch tests to equivalent MSQ tests in the new IT Framework. A new MSQ IT test was added which will pick up any json files present in `integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index`.

Also increased the verbosity of the tests running in the new IT framework.

### CODE REVIEW

File: integration-tests/src/main/java/org/apache/druid/testing/utils/MsqTestQueryHelper.java [hunk 1]
```diff
@@ -95,7 +95,7 @@ public String getQueryURL(String schemeAndHost)
    */
   public SqlTaskStatus submitMsqTask(String sqlQueryString) throws ExecutionException, InterruptedException
   {
-    return submitMsqTask(new SqlQuery(sqlQueryString, null, false, false, false, ImmutableMap.of(), null));
+    return submitMsqTask(new SqlQuery(sqlQueryString, null, false, false, false, ImmutableMap.of("finalizeAggregations", false), null));
```
Comment (cryptoe) [inline integration-tests/src/main/java/org/apache/druid/testing/utils/MsqTestQueryHelper.java:98]: Long fix would be change it to a sql query builder. 
The short fix would be to have the context param as part of the function argument. 
I donot think its correct to have "finalizeAggregations", false as default

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITMSQForBatchIndex.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      LOG.info("Sql Task submitted with task Id - %s", sqlTaskStatus.getTaskId());
+
+      if (sqlTaskStatus.getState().isFailure()) {
+        Assert.fail(StringUtils.format(
+              "Unable to start the task successfully.\nPossible exception: %s",
+              sqlTaskStatus.getError()
+        ));
+      }
+      msqHelper.pollTaskIdForCompletion(sqlTaskStatus.getTaskId());
+      dataLoaderHelper.waitUntilDatasourceIsReady(datasource + files[i].getName());
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITMSQForBatchIndex.java:139]: I think this should be pushed inside msqHelper or a base test class. It would most likely get copied in every test case.

File: integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index/wikipedia_index_msq.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
...bot\",\"type\":\"string\"},{\"name\":\"anonymous\",\"type\":\"string\"},{\"name\":\"namespace\",\"type\":\"string\"},{\"name\":\"continent\",\"type\":\"string\"},{\"name\":\"country\",\"type\":\"string\"},{\"name\":\"region\",\"type\":\"string\"},{\"name\":\"city\",\"type\":\"string\"},{\"name\":\"added\",\"type\":\"double\"},{\"name\":\"deleted\",\"type\":\"double\"},{\"name\":\"delta\",\"type\":\"double\"}]'\n  )\n))\nSELECT\n  TIME_FLOOR(CASE WHEN CAST(\"timestamp\" AS BIGINT) > 0 THEN MILLIS_TO_TIMESTAMP(CAST(\"timestamp\" AS BIGINT)) ELSE TIME_PARSE(\"timestamp\") END, 'PT1S') AS __time,\n  \"page\",\n  \"language\",\n  \"user\",\n  \"unpatrolled\",\n  \"newPage\",\n  \"robot\",\n  \"anonymous\",\n  \"namespace\",\n  \"continent\",\n  \"country\",\n  \"region\",\n  \"city\",\n  COUNT(*) AS \"count\",\n  SUM(\"added\") AS \"added\",\n  SUM(\"deleted\") AS \"deleted\",\n  SUM(\"delta\") AS \"delta\",\n  APPROX_COUNT_DISTINCT_DS_THETA(\"user\") AS \"thetaSketch\",\n  DS_QUANTILES_SKETCH(\"delta\") AS \"quantilesDoublesSketch\",\n  APPROX_COUNT_DISTINCT_DS_HLL(\"user\") AS \"HLLSketchBuild\"\nFROM \"source\"\nGROUP BY 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13\nPARTITIONED BY DAY",
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index/wikipedia_index_msq.json:2]: Is it possible to prettify this long str somehow ?

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITMSQForBatchIndex.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+@RunWith(DruidTestRunner.class)
+@Category(MultiStageQuery.class)
+public class ITMSQForBatchIndex
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITMSQForBatchIndex.java:52]: NIt: ITSQLBasedBatchIngestion

File: integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index/wikipedia_index_msq.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+  "testQueriesFilePath" : "/multi-stage-query/batch-index/test-queries/wikipedia_index_msq_queries.json"
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index/wikipedia_index_msq.json:3]: So if every test case has the same q's then it makes sense to create a new file else I would add the native q's here directly.

File: integration-tests/src/main/java/org/apache/druid/testing/utils/MsqSqlTasksWithTestQueries.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class MsqSqlTasksWithTestQueries
+{
+  private final String sqlTask;
+  private final String testQueriesFilePath;
+
+  @JsonCreator
+  public MsqSqlTasksWithTestQueries(
```
Comment (cryptoe) [inline integration-tests/src/main/java/org/apache/druid/testing/utils/MsqSqlTasksWithTestQueries.java:31]: I think this class is not needed. 
checkout : ITMultiStageQuery

**Review by cryptoe — COMMENTED** [review]
@abhagraw Thanks for the PR. Left an initial review.

File: integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index/wikipedia_index_msq.json (hunk 3 above)
Comment (abhagraw) [reply integration-tests-ex/cases/src/test/resources/multi-stage-query/batch-index/wikipedia_index_msq.json]: json files does not support pretty multi-line string syntax.
ref - https://stackoverflow.com/a/52557585

Have changed structure, and not MSQ sql ingestion queries are placed in separate sql files which are more readable.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  @Test
+  public void testSQLBasedBatchIngestion() throws Exception
+  {
+    // Get list of all directories in batch-index folder. Each folder is considered a test case.
+    File[] directories = (new File(getClass().getResource(BATCH_INDEX_TASKS_DIR).toURI())).listFiles(File::isDirectory);
+    int fail_count = 0;
+    LOG.info("Test will be run for sql in the following directories - \n %s", Arrays.toString(directories));
+    for (File dir : directories) {
+      try {
+        runMSQTaskandTestQueries(BATCH_INDEX_TASKS_DIR + dir.getName(), "msqBatchIndex_" + dir.getName(),
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java:47]: Can we create a parameterized test?
That way we will know which test is exactly flaky

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  protected String getFileWithFormatFromDir(String dir, String format) throws URISyntaxException
+  {
+    File[] file = (new File(getClass().getResource(dir).toURI())).listFiles(pathname -> {
+      String name = pathname.getName();
+      return name.endsWith(format) && pathname.isFile();
+    });
+    return dir + '/' + file[0].getName();
+  }
+
+  protected String getSqlStringFromDir(String dir, String datasource) throws URISyntaxException
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java:60]: We are doing the replacement for the data source. So we should mention that in the signature. 
Lets javadoc this method.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      queryHelper.testQueriesFromString(queryResponseTemplate);
+
+    }
+    catch (Exception e) {
+      LOG.error(e, "Error while running test query");
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected void runMSQTaskandTestQueries(String relativePath, String datasource, Map<String, Object> msqContext) throws Exception
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java:108]: lets java doc this method.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    sqlString = StringUtils.replace(
+        sqlString,
+        "%%DATASOURCE%%",
+        datasource
+    );
+
+    return sqlString;
+  }
+
+  protected void doTestQuery(String dir, String dataSource) throws URISyntaxException
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java:81]: this as well since it will be used in all tests.

File: distribution/pom.xml [hunk 11]
```diff
@@ -225,6 +225,8 @@
                                         <argument>-c</argument>
                                         <argument>org.apache.druid.extensions:druid-multi-stage-query</argument>
                                         <argument>-c</argument>
+                                        <argument>org.apache.druid.extensions:druid-catalog</argument>
```
Comment (LakshSingla) [inline distribution/pom.xml:228]: Is this change related to this PR? If not, then should this be raised as a separate PR?

File: integration-tests-ex/cases/pom.xml [hunk 12]
```diff
@@ -346,7 +351,7 @@
                             </goals>
                             <configuration>
                                 <!-- our tests are very verbose, let's keep the volume down -->
-                                <redirectTestOutputToFile>true</redirectTestOutputToFile>
+                                <redirectTestOutputToFile>False</redirectTestOutputToFile>
```
Comment (LakshSingla) [inline integration-tests-ex/cases/pom.xml:354]: Why is this change required? Also, if we go ahead with the change, then the comment above this should be updated to explain why it is set to false
```suggestion
                                <redirectTestOutputToFile>false</redirectTestOutputToFile>
```

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+  @Inject
+  private MsqTestQueryHelper msqHelper;
+
+  @Inject
+  protected TestQueryHelper queryHelper;
+
+  @Inject
+  private DataLoaderHelper dataLoaderHelper;
+
+  protected String getFileWithFormatFromDir(String dir, String format) throws URISyntaxException
```
Comment (LakshSingla) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/AbstractITSQLBasedBatchIngestion.java:51]: What's the advantage of the method if it is returning just the first file ending with `format`. Should the user pass the filename himself?

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java (hunk 7 above)
Comment (LakshSingla) [reply integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java:47]: I agree with @cryptoe. Either we should pass the filename individually in each test to a standard helper method or create parameterized tests. Extracting filenames from the dir in a test makes it harder to debug failing tests, and also are difficult to pinpoint why the tests are failing clearly.

Comment (abhagraw) [reply integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java]: Have parameterized the tests now. All the SQL ingestion task and test queries have been moved in the same folder.
A testcase would a be a pair of SQL ingestion task filename and test queries filename.

File: distribution/pom.xml (hunk 11 above)
Comment (abhagraw) [reply distribution/pom.xml:228]: The docker containers created for the test fail with the following error, if this extension is not added - 
`Listening for transport dt_socket at address: 8000
2022-11-15T06:35:58,107 INFO [main] org.hibernate.validator.internal.util.Version - HV000001: Hibernate Validator 5.2.5.Final
Exception in thread "main" org.apache.druid.java.util.common.ISE: Extension [/usr/local/apache-druid-2022.12.0-iap-SNAPSHOT/extensions/druid-catalog] specified in "druid.extensions.loadList" didn't exist!?
	at org.apache.druid.guice.ExtensionsLoader.getExtensionFilesToLoad(ExtensionsLoader.java:183)
	at org.apache.druid.guice.ExtensionsLoader$ServiceLoadingFromExtensions.addAllFromFileSystem(ExtensionsLoader.java:292)
	at org.apache.druid.guice.ExtensionsLoader$ServiceLoadingFromExtensions.<init>(ExtensionsLoader.java:280)
	at org.apache.druid.guice.ExtensionsLoader$ServiceLoadingFromExtensions.<init>(ExtensionsLoader.java:268)
	at org.apache.druid.guice.ExtensionsLoader.lambda$getFromExtensions$0(ExtensionsLoader.java:139)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at org.apache.druid.guice.ExtensionsLoader.getFromExtensions(ExtensionsLoader.java:137)
	at org.apache.druid.cli.Main.main(Main.java:98)`

File: integration-tests-ex/cases/pom.xml (hunk 12 above)
Comment (abhagraw) [reply integration-tests-ex/cases/pom.xml:354]: Updated the comment to explain the change.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    try {
+      runMSQTaskandTestQueries(BATCH_INDEX_TASKS_DIR + sqlFileName,
+                               BATCH_INDEX_TASKS_DIR + queryFileName,
+                               FilenameUtils.removeExtension(sqlFileName),
+                               ImmutableMap.of("finalizeAggregations", false,
+                                               "maxNumTasks", 10,
+                                               "groupByEnableMultiValueUnnesting", false
+                               ));
+    }
+    catch (Exception e) {
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java:66]: We might have to throw exception here. 
Please log the current sql file and the query file name also.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  @Test
+  @Parameters(method = "test_cases")
+  public void testSQLBasedBatchIngestion(String sqlFileName, String queryFileName)
+  {
+    try {
+      runMSQTaskandTestQueries(BATCH_INDEX_TASKS_DIR + sqlFileName,
+                               BATCH_INDEX_TASKS_DIR + queryFileName,
+                               FilenameUtils.removeExtension(sqlFileName),
+                               ImmutableMap.of("finalizeAggregations", false,
+                                               "maxNumTasks", 10,
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java:62]: Should this be 3 ?  I donot think we need so many slots for IT's

**Review by cryptoe — COMMENTED** [review]
Some comments. 
LGTM post they are addressed and CICD is 🟢

**Comment by cryptoe** [comment]
Another nit we should fix in this PR : 
```
2022-11-18T19:26:19,074 INFO [main] org.apache.druid.testing.utils.TestQueryHelper - Sql Task submitted with task Id - query-77b14a7e-999d-4756-92eb-9fba59f68836

2022-11-18T19:26:19,105 WARN [main] org.apache.druid.java.util.common.RetryUtils - Retrying (1 of 99) in 895ms.

org.apache.druid.testing.utils.MsqTestQueryHelper$TaskStillRunningException: null

	at org.apache.druid.testing.utils.MsqTestQueryHelper.lambda$pollTaskIdForCompletion$0(MsqTestQueryHelper.java:163) ~[druid-integration-tests-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]

	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:129) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]

	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:81) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]

	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:163) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]

	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:153) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]
	
	```
	
	```
	public TaskState pollTaskIdForCompletion(String taskId) throws Exception
  {
    return RetryUtils.retry(
        () -> {
          TaskStatusPlus taskStatusPlus = overlordClient.getTaskStatus(taskId);
          TaskState statusCode = taskStatusPlus.getStatusCode();
          if (statusCode != null && statusCode.isComplete()) {
            return taskStatusPlus.getStatusCode();
          }
          throw new TaskStillRunningException();
        },
        (Throwable t) -> t instanceof TaskStillRunningException,
        100
    );
  }
  ```
	let's change this method to not throw an exception but just log unable to get the status of the task.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java (hunk 15 above)
Comment (abhagraw) [reply integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/ITSQLBasedBatchIngestion.java]: Reduced the maxNumTasks to 5

**Comment by abhagraw** [comment]
> Another nit we should fix in this PR :
> 
> ```
> 2022-11-18T19:26:19,074 INFO [main] org.apache.druid.testing.utils.TestQueryHelper - Sql Task submitted with task Id - query-77b14a7e-999d-4756-92eb-9fba59f68836
> 
> 2022-11-18T19:26:19,105 WARN [main] org.apache.druid.java.util.common.RetryUtils - Retrying (1 of 99) in 895ms.
> 
> org.apache.druid.testing.utils.MsqTestQueryHelper$TaskStillRunningException: null
> 
> 	at org.apache.druid.testing.utils.MsqTestQueryHelper.lambda$pollTaskIdForCompletion$0(MsqTestQueryHelper.java:163) ~[druid-integration-tests-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]
> 
> 	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:129) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]
> 
> 	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:81) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]
> 
> 	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:163) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]
> 
> 	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:153) ~[druid-core-25.0.0-SNAPSHOT.jar:25.0.0-SNAPSHOT]
> 	
> ```
> 
> ```java
> public TaskState pollTaskIdForCompletion(String taskId) throws Exception
>   {
>     return RetryUtils.retry(
>         () -> {
>           TaskStatusPlus taskStatusPlus = overlordClient.getTaskStatus(taskId);
>           TaskState statusCode = taskStatusPlus.getStatusCode();
>           if (statusCode != null && statusCode.isComplete()) {
>             return taskStatusPlus.getStatusCode();
>           }
>           throw new TaskStillRunningException();
>         },
>         (Throwable t) -> t instanceof TaskStillRunningException,
>         100
>     );
>   }
> ```
> 
> Let's change this method to not throw an exception but just log unable to get the status of the task. We can do this as part of another PR as well.

Used quite retry - so that not all retries throw the exception + stacktrace.

**Review by cryptoe — APPROVED** [review]
LGTM. 
Thanks for the contribution @abhagraw
